sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -568,7 +568,14 @@ public virtual Guid ParseGuidBytes(ReadOnlyMemory<byte> bytes)
         {
             if (bytes.Length == GuidSizeInBytes)
             {
-                return new Guid(bytes.Span.ToArray());
+                // Use TryRead to avoid allocating an array
+                if (!MemoryMarshal.TryRead<Guid>(bytes.Span, out var lockTokenGuid))
```

Comment:
The direct use of span makes sense. Having said that, I started to wonder about endianness here. Guid struct has 1 Int32 and 2 Int16 fields. TryRead will work only if the passed in bytes are the right endianness. Is that the case here? i.e. what's the source of these bytes?

Reply:
The bytes are sent over the wire by the Service Bus service.

Reply:
I think it uses [this part ](https://github.com/Azure/azure-amqp/blob/0c4cf8a15b7df2cbe155e59c26ea23cfb6d3f652/src/Encoding/AmqpBitConverter.cs#L374)for the delivery tag but I'm not an expert

Reply:
> TryRead will work only if the passed in bytes are the right endianness

Why is this specific to TryRead? Wouldn't there be the same issue when using the ctor?

Reply:
Which would confirm your hunch @KrzysztofCwalina because that encodes big endian and reads big endian in the read methods. So we would always hit the ToArray case 😭

Reply:
Does TryRead not working mean it will return false? Or give the incorrect value?

Reply:
> Wouldn't there be the same issue when using the ctor?

@JoshLove-msft, the same issue does exist for the ctor, see the branches in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Guid.cs#L49

Reply:
This should be verifiable by debugging one of the live tests right?

Reply:
So I assume it is working correctly as this code has been around for years. Even if it somehow reversed the bits it would still probably "work" as the lock token acts as an in-memory handle to the AMQP Delivery.

Reply:
It happens to "work" because in the normal course of operation, the client does not care about the interpretation of the bytes. They get transported from the server, TryRead into Guid, and then probably bit-blasted as-is when the delivery handle is sent back to the server. 

But, it might be a problem for distributed tracing, i.e. when the server logs the GUID as string, it might be different string value than what's logged on the server. 

BTW, I also noticed that the server code assumes the code runs on a little-endian machine: https://github.com/Azure/azure-amqp/blob/0c4cf8a15b7df2cbe155e59c26ea23cfb6d3f652/src/Encoding/AmqpBitConverter.cs#L374. For the code to be portable across little-big endian architectures, it should probably have a similar branch to what's inside the Guid's ctor in .NET.

Reply:
> Which would confirm your hunch @KrzysztofCwalina because that encodes big endian and reads big endian in the read methods. So we would always hit the ToArray case 😭

My understanding is that TryRead should only be used for little endian, not that it would return false on big endian. This means that the current code could end up reversing the lock token if the client machine is actually big endian. Using the Guid constructor handles both cases at the cost of us needing to allocate always. We can do an endian check ourselves to only use TryRead for little endian to avoid the extra allocation.

Re: the AMQP lib handling that could cause issues now that we no longer have full control over the machine where the ReceivedMessage is serialized. Since we are serializing the received message as part of this PR, we should probably file an issue so that the AMQP lib has handling for endianness.

Reply:
https://github.com/Azure/azure-amqp/issues/240

Reply:
Sorry to chime into the party again. Please let me know when I'm getting annoying. I'm pretty much here just for my learning experience.

Could it be that we are making false assumptions?

By looking at the Transfer it looks like the DeliveryTag is encoded binary

https://github.com/danielmarbach/azure-amqp/blob/b26e5c701fd82142d29ff8d9588a387bc61c6db2/src/Framing/Transfer.cs#L123

it is also decoded binary.

Reply:
Your feedback is always appreciated @danielmarbach! 

> By looking at the Transfer it looks like the DeliveryTag is encoded binary

Interesting, from a spec perspective the [delivery tag](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-delivery-tag) is just up to 32 opaque bytes. 

I think the 16 bytes GUID is an implementation detail of Service Bus. But still the issue about the encoding of GUID not being endianness aware seems to be something that should be fixed.

Reply:
When you look at [this fiddle](https://dotnetfiddle.net/2QINUw) it shows the code that is already "in production" would horribly fail today since the Guid encoding would not be preserved.

The AmqpSender also uses the DeliveryTag to send an incrementing batch number. This shows the delivery tag is "arbitrary bytes" right? When the service sends us the lock token guid over the delivery tag it comes through as binary encoded and the AMQP lib will not do any sort of interpretation if my the code I pointed out is the right one.

Reply:
> When you look at [this fiddle](https://dotnetfiddle.net/2QINUw) it shows the code that is already "in production" would horribly fail today since the Guid encoding would not be preserved.

I think we are in agreement - the data is encoded as arbitrary bytes as per the spec. However, the service is constructing these bytes from a Guid, and that is why the library is parsing into a Guid.

Reply:
> But still the issue about the encoding of GUID not being endianness aware seems to be something that should be fixed.

Correct but I believe this is not a problem in this code because like you said and I hinted at above the DeliveryTag is opaque for the AMQP lib and it will just send or receive whether the client or service brings. 

That being said, the original version which was tweaked by me months ago used Buffer.BlockCopy which deals with endianness.

Reply:
Yep, we still need the endianness handling because otherwise our LockToken property could appear incorrect on big endian machines.

Reply:
What does the service send? When calling ToByteArray on GUID you essentially get "mixed-endian"

https://stackoverflow.com/questions/10190817/guid-byte-order-in-net?utm_source=pocket_saves

Reply:
By the way that bug then also exists in https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs#L479-L482

and to properly address this you essentially have to reimplement the Guid constructor @KrzysztofCwalina mentioned and also TryWriteBytes  on the TFMs that don't support it or go back to use BufferBlockCopy or ToByteArray which would deal with endianness.

My changes have been merged almost two years ago. I wonder why we haven't heard more about this problem. Probably because big endian machines are so seldom used?

Reply:
> My changes have been merged almost two years ago. I wonder why we haven't heard more about this problem. Probably because big endian machines are so seldom used?

I suspect that is the reason. I'm going to file an issue to investigate/address all Guid related serialization/deserialization issues.

Reply:
![](https://media.giphy.com/media/RWUqVYucDBD4A/giphy.gif)

Reply:
No worries at all - we reviewed these PRs and missed it too.

Reply:
Created https://github.com/Azure/azure-sdk-for-net/issues/33825

Reply:
@jsquire we I did the hash code optimization in Event Hubs you somehow discovered the endianness problems on some test runs. Is event hubs using different agents than Service Bus? Could Service Bus also leverage a similar mechanism to test on more exotic operating systems / agents, like event hub does (assuming it does)?

Reply:
I noticed it running on macOS ARM (big endian) - which was coincidental because it happens to be the machine that MS issued me.   Unfortunately, the Azure DevOps machines do not have M1 specs available at this point so there's no way that I'm aware of to automate it.

I'm happy to help by running snippets as a test in the meantime, if we want to confirm behavior.